File: Maximum Matrix Sum/Maximum Matrix Sum.py
# Solution for Maximum Matrix Sum in PY

class Solution:
    def maxMatrixSum(self, matrix) -> int:

        # Total sum of absolute values
        total_absolute_sum = 0

        # Smallest absolute value in the matrix
        minimum_absolute_value = float('inf')

        # Count of negative numbers
        negative_count = 0

        matrix_size = len(matrix)

        # Traverse the matrix
        for row in range(matrix_size):
            for col in range(matrix_size):

                current_value = matrix[row][col]

                # Add absolute value
                total_absolute_sum += abs(current_value)

                # Count negatives
                if current_value < 0:
                    negative_count += 1

                # Track minimum absolute value
                minimum_absolute_value = min(
                    minimum_absolute_value, abs(current_value)
                )

        # Apply logic based on negative count parity
        if negative_count % 2 == 0:
            return total_absolute_sum

        return total_absolute_sum - 2 * minimum_absolute_value

File: Maximum Matrix Sum/Maximum Matrix Sum.java
// Solution for Maximum Matrix Sum in JAVA

class Solution {
    public long maxMatrixSum(int[][] matrix) {

        // Stores the total sum of absolute values of all elements
        long totalAbsoluteSum = 0;

        // Tracks the smallest absolute value in the matrix
        int minimumAbsoluteValue = Integer.MAX_VALUE;

        // Counts the number of negative elements
        int negativeCount = 0;

        int matrixSize = matrix.length;

        // Traverse the entire matrix
        for (int row = 0; row < matrixSize; row++) {
            for (int col = 0; col < matrixSize; col++) {

                int currentValue = matrix[row][col];

                // Add absolute value to total sum
                totalAbsoluteSum += Math.abs(currentValue);

                // Count negatives
                if (currentValue < 0) {
                    negativeCount++;
                }

                // Track minimum absolute value
                minimumAbsoluteValue =
                        Math.min(minimumAbsoluteValue, Math.abs(currentValue));
            }
        }

        /*
            If number of negative values is even:
            → All values can be made positive
            → Maximum sum = total absolute sum

            If odd:
            → One value must stay negative
            → Subtract 2 * smallest absolute value
        */
        if (negativeCount % 2 == 0) {
            return totalAbsoluteSum;
        }

        return totalAbsoluteSum - 2L * minimumAbsoluteValue;
    }
}
